### PR TITLE
sidecars that react to configmap changes

### PR DESCRIPTION
**Pull request by jimethn**
Instead of filling out the directories and seeding the trust stores once during startup (via initContainers), instead keep sidecars around that watch for configMaps to be added/removed and automatically update the directories. This required updating to a newer version of k8s-sidecar that had the watch mode working with binaryData.

To handle updating the user & admin trust stores, simply remove the trust store persistence and allow SEED_ON_STARTUP. When a sidecar detects a change, it kills the mtls process, causing it to start back up and re-seed the trust store. This solution removes the need to handle trusted key removal in the sidecar script.

### CODE REVIEW

File: manifests/configmap-seed/remove-persistence-pvc.patch.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+metadata:
+  name: mtls
+spec:
+  template:
+    spec:
+      containers:
+        - name: mtls
+          volumeMounts:
+            - $patch: delete
+              mountPath: /home/mtls/database
```
Comment (drGrove) [inline manifests/configmap-seed/remove-persistence-pvc.patch.yml]: Why delete the database?

Comment (jimethn) [reply manifests/configmap-seed/remove-persistence-pvc.patch.yml]: I thought the DB was used for storing which GPG keys are trusted?

Comment (drGrove) [reply manifests/configmap-seed/remove-persistence-pvc.patch.yml]: No, the DB is for certificate reference

File: manifests/configmap-seed/kustomization.yml [hunk 2]
```diff
@@ -5,13 +5,12 @@ resources:
...
+  - remove-persistence-pvc.patch.yml
 configMapGenerator:
-  - name: mtls-seed-script
+  - name: mtls-restart-script
     files:
-      - ./files/seed.sh
+      - ./files/restart-mtls.sh
 images:
   - name: kiwigrid/k8s-sidecar
-    newName: drgrove/k8s-sidecar
```
Comment (james-callahan) [inline manifests/configmap-seed/kustomization.yml:16]: Add a `newTag` here as documentation of what the shasum belongs to.

File: manifests/configmap-seed/kustomization.yml [hunk 3]
```diff
@@ -5,13 +5,12 @@ resources:
...
-    digest: sha256:8799b42ad45245b63a85e0e2190914da8d32472f3cd98f057cc332ca32be98cc
+    digest: sha256:fcb3ab84daf0269e811cffa5334e3acb75abd215
```
Comment (james-callahan) [inline manifests/configmap-seed/kustomization.yml]: this shasum looks like it got cut off

**Review by james-callahan — CHANGES_REQUESTED** [review]

File: manifests/configmap-seed/files/restart-mtls.sh [hunk 4]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+echo "Restarting mtls process in main container"
+
+killall uwsgi
```
Comment (drGrove) [inline manifests/configmap-seed/files/restart-mtls.sh:5]: The start-up time on mtls can be rather large. we should look at doing this as part of ether the main service or via a sidecar.

File: manifests/configmap-seed/kustomization.yml (hunk 3 above)
Comment (jimethn) [reply manifests/configmap-seed/kustomization.yml]: Fixed

File: manifests/configmap-seed/remove-persistence-pvc.patch.yml (hunk 1 above)
Comment (jimethn) [reply manifests/configmap-seed/remove-persistence-pvc.patch.yml]: But isn't there trust data stored in there along with the certificates? I'm looking at e.g. [verify()](https://github.com/drGrove/mtls-server/blob/ea3d8f5179d91966ef4c488ccb078ee0f915a86c/mtls_server/cert_processor.py#L97) and [add_user()](https://github.com/drGrove/mtls-server/blob/ea3d8f5179d91966ef4c488ccb078ee0f915a86c/mtls_server/handler.py#L173).

The premise of this PR is that we get easy user management by just being ephemeral. If a user is removed, but their certificates are still in the DB, then that gives them an unintended level of access, right? Or am I confused?

File: manifests/configmap-seed/files/restart-mtls.sh (hunk 4 above)
Comment (jimethn) [reply manifests/configmap-seed/files/restart-mtls.sh:5]: How large is large? Maybe it's acceptable for the use case. It only needs to be available the moment someone requests a cert, right? And that happens maybe once a day given the 18h timeout.

Comment (jimethn) [reply manifests/configmap-seed/files/restart-mtls.sh:5]: Consider, also, the problem I'm trying to solve here: Currently, if the user list is updated, the Pod must be manually restarted in order to update the server. With this change, the Pod restarts itself. It's objectively better, even if it's not perfect.

File: manifests/configmap-seed/remove-persistence-pvc.patch.yml (hunk 1 above)
Comment (drGrove) [reply manifests/configmap-seed/remove-persistence-pvc.patch.yml]: GPG signature is what is determined for authentication, if their certificate is left over they will still not be able to access the service as long as their key has been removed. So they will not receive a new certificate

**Comment by drGrove** [comment]
Please rebase as this is now out of date

File: manifests/configmap-seed/remove-persistence-pvc.patch.yml (hunk 1 above)
Comment (jimethn) [reply manifests/configmap-seed/remove-persistence-pvc.patch.yml]: Thanks for explaining. Database: kept.

**Comment by jimethn** [comment]
Rebased and squashed

**Review by drGrove — APPROVED** [review]

**Pull request merged**